### ISSUE TITLE
Master CLI gets stuck after running for 10-30 Minutes

### ISSUE THREAD

**Issue by hujiko** [haproxy/haproxy#1512]
### Detailed Description of the Problem

I am using a Phython-Script that automatically connects to the Master CLI Unix socket, executing
"show proc" and then for every worker "@!<pid> show stat".

I use that to get statistical information (once per Minute) about the currently active, but also about previous workers that still hold connections.

It was working fine in version 2.5-dev2, when a commit was introducted to support this for stopped workers.
Now I upgraded to 2.5.0-1ppa on Ubuntu 20 and after 10-30 Minutes of runtime, the Master-CLI seems to be stuck.

When I connect to it even manually with `nc -U /var/run/haproxy-master.sock` and execute `show proc`, I don't get any response.

Interestingly, triggering a HAProxy reload seems to fix the issue, until it re-appears after 10-30 Minutes.

### Expected Behavior

I would expect, that the socket replies to commands constantly

### Steps to Reproduce the Behavior

- Start HAProxy in master-worker mode and with `-S /run/haproxy-master.sock`
- Handle a few hundred/thousand requests through HAProxy
- Connect to the master-CLI socket: `nc -U /var/run/haproxy-master.sock`
- Try to list current processes: `show proc`
- There will be no response

Interestingly, this is not reproducible for me on an instance without much traffic, but it also is not reproduceable on an instance with a lot of traffic, when I disable the regular fetching of metrics. But there I am still testing.

### Do you have any idea what may have caused this?

Not really.
I guess it must be something that has changed between this commit 07954fb069828e311cef9b2bc54a5b4deed96508 and the final version 2.5.0 in regards to how the something about master-CLI socket is handled. 

### What is your configuration?

```haproxy
global
  log 127.0.0.1:9001 local0 notice
  log 127.0.0.1:9001 local1 info
  maxconn 100000
  ulimit-n 655360

  external-check
  spread-checks 10
  insecure-fork-wanted

  master-worker

  stats socket /var/run/haproxy/admin.sock mode 666 expose-fd listeners level admin
  stats socket /var/run/haproxy/haproxy.sock expose-fd listeners mode 666 level user
  stats timeout 30s
  user haproxy
  group haproxy
  daemon

  # Generated by https://ssl-config.mozilla.org/#server=haproxy&server-version=1.8.8&config=intermediate
  ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
  ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

  ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
  ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

  tune.ssl.default-dh-param 2048
  tune.bufsize 524288

defaults
  log global
  option  dontlognull
  option  redispatch
  option  log-health-checks
  timeout connect 5s
  timeout client  600s
  timeout server  600s
  timeout check   250
  errorfile 400 /etc/haproxy/errors/400.http
  errorfile 403 /etc/haproxy/errors/403.http
  errorfile 408 /etc/haproxy/errors/408.http
  errorfile 500 /etc/haproxy/errors/500.http
  errorfile 502 /etc/haproxy/errors/502.http
  errorfile 503 /etc/haproxy/errors/503.http
  errorfile 504 /etc/haproxy/errors/504.http
  default-server maxconn 100000
  option allbackups

listen stats # Define a listen section called "stats"
  bind 10.0.0.10:443  ssl crt /etc/ssl/haproxy/ alpn http/1.1
  bind 10.0.0.10:80
  mode http

  http-request redirect scheme https unless { ssl_fc }

  stats enable  # Enable stats page
  stats hide-version  # Hide HAProxy version
  stats uri /  # Stats URI
  # stats admin if TRUE
  log                  global
  option               httplog
  option               dontlognull
  maxconn              50

listen alive_check
  bind 127.0.0.1:8080
  mode http
  http-request return status 200 content-type text/plain string "ALIVE"

program reload_notify
  command /bin/bash -c "/etc/haproxy/reload_notify.sh"

listen DB_EXAMPLE
  mode tcp
  bind 10.0.0.3:3306
  option tcplog
  timeout client 8h
  timeout server 8h

    option httpchk
  http-check expect status 200
  http-check expect string YES
  timeout check 500ms
  default-server inter 5s


  server hostname 10.0.0.4:3306  check port 81

listen HTTP_EXAMPLE
  mode http
  bind 10.0.0.1:80
  bind 10.0.0.1:443 ssl alpn h2,http/1.1 crt /etc/ssl/haproxy/
  option httpchk GET /
  option  httplog
  option forwardfor header X-Real-Ip

  http-request del-header X-Forwarded-For
  http-request del-header X-Forwarded-Port
  http-request del-header X-Forwarded-Proto
  http-request set-header X-Forwarded-Proto https if { ssl_fc }
  http-request set-header X-Forwarded-Proto http if !{ ssl_fc }


  timeout check 1500ms
  default-server inter 2s

  http-response set-header X-LB-FQDN "FOOBAR"
  http-response set-header X-LB-Backend-Server %s
  # https://serverfault.com/questions/650588/haproxy-timing-connection-diagram
  http-response set-header X-LB-Times "connect: %Th ms; queue: %Tw ms; be-connect: %Tc ms; be-response: %Tr ms"

  http-request redirect scheme https code 301 unless { ssl_fc }

  server mymachine 10.0.0.2:80 check

(The actual config is much longer and has ~120 listen blocks)
In case the full config is neccessary, I am happy to provide it. Just replacing confidential data in such a huge config will take some time.
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.5.0-1ppa1~focal 2021/11/26 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.0.html
Running on: Linux 5.4.0-80-generic #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-IUgeIz/haproxy-2.5.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
No errors are being logged
```

### Additional Information

`uname -a`

```
Linux <hostname> 5.4.0-80-generic #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by wlallemand** [comment]
Just to be sure, are you sure you are reaching the right path, because I see several UNIX sockets and that's confusing:

> 
>     * Start HAProxy in master-worker mode and with `-S /run/haproxy-master.sock`

This is not the same path as the one above

>     * Connect to the master-CLI socket: `nc -U /var/run/haproxy-master.sock`


> program reload_notify
>   command /bin/bash -c "/etc/haproxy/reload_notify.sh"

I don't know if it's related to your problem, probably not but it looks like an ugly hack, what are you trying to do here?

**Comment by hujiko** [comment]
> ls -lha run
> lrwxrwxrwx 1 root root 4 Jun 20  2021 run -> /run

Good point.
I checked:
```
ls -lha /var/run
lrwxrwxrwx 1 root root 4 Jun 20  2021 /var/run -> /run
```
That's a symlink, so it ends up in the same file.

> I don't know if it's related to your problem, probably not but it looks like an ugly hack, what are you trying to do here?

I want to execute a script, whenever HAProxy performs a reload, so that I can store that information in a monitoring tool.

**Comment by hujiko** [comment]
I still fail to reproduce it outside of our production environment.
I guess it has something to to with having loads of open connections simultanously and also constantly reading from the Master-CLI

But I was able to narrow it down.
The Bug for me was introduced between v2.5-dev6 and v2.5-dev7.
See diff here: https://github.com/haproxy/haproxy/compare/v2.5-dev6...v2.5-dev7

I saw some code changes regarding variable locks, but am not skilled enough to fully understand it.

Currently my suspicion is, that a variable gets locked, but never unlocked. And therefore the "show proc" is waiting forever for the unlock.

But as a reload fixes the issue, even though the old process stays alive (as it still holds connections), the reload-operation seems to unlock that variable.

Does that make sense?

**Comment by wlallemand** [comment]
I don't see anything related to this in the changelog. And unfortunately no, we don't use any lock for this feature.

However, I might have an idea about what's going on, I don't use `nc -U` often and it looks like it doesn't close the connection by default, like `socat` does, and no timeout is set on the master CLI. So maybe you just exhausted the maxconn of the master CLI which is 10. Maybe you should check if your nc processes left.

**Comment by wtarreau** [comment]
@hujiko locks held will block all other uses of the same lock, and haproxy is very strict on this, it waits no more than 2 seconds before killing the process (you never want your LB to stay up but deadlocked in production).
So it must not be related. Also if you're only sending "show proc" it doesn't involve variables either.

Potential candidates include the various fixes on the channels (end of stream etc), but for some of them there were also a few more fixes merged into 2.5.1. Would you please give it a try so that we don't all waste time possibly trying to debug something that was recently fixed ?

Thanks!

**Comment by hujiko** [comment]
> Maybe you should check if your nc processes left.

Good catch. But sadly this was not my issue. I checked and there were not processes running any more that could potentially still hold a connection to that socket.

> Would you please give it a try so that we don't all waste time possibly trying to debug something that was recently fixed ?

I tried compiling everything up to this commit: aab8d255bc0fcbcc50884a4be4f69598ee08fe73 (which is from yesterday)
but the issue still occurs.

By going through the git history, compiling versions and test them, I was able to trace it down to this commit:
https://github.com/haproxy/haproxy/commit/b7308f00cb03e5accd2f6092bf09fc2ac1360ff9

The build including this commit produces the issue I reported. Compiling the previous commit, everything works fine.

**Comment by wtarreau** [comment]
Thanks for this! That's really not cool. So basically we have the choice between two bugs. We'll need to sit down around a table and analyse them in depth, we can't continuously fix and revert patches to satisfy only half of the users at a time :-(

CC @capflam

**Comment by capflam** [comment]
Hum, that's annoying because this commit reverts some changes to go back to a solution used since a long time. So I'm surprised it blocks the CLI. We must first check if it is not a side effect.

**Comment by hujiko** [comment]
Totally agree. Reverting the revert is not a solution.

If you need/want me to apply any kind of patch to see, if it solves my problem: I am very happy to do that.
My knowledge about stream handling is too limited to properly understand, what this code is doing :/

**Comment by wtarreau** [comment]
Anyway we need to consider all this as a whole. The fix you attempted Christopher was valid but broke other issues. In parallel there's abortonclose which only partially works nowadays as reported in #1505. I think it's interesting that we have several corner cases to consider together to see if we can design a better and more robust solution to all these problems of shutdowns and aborts.

**Comment by capflam** [comment]
I found a bug on the CLI that may explain your issue. If the script requesting the master CLI sends a partial command (or any data except newline or semicolon) before closing, the connection remains opened on HAProxy side. The `maxconn` is set to 10 and there is no timeout. Thus, the CLI socket can be quickly unresponsive in this case. Here is a patch to fix this bug. It may fix your issue.
I will wait for @wlallemand review before pushing it. Could you try it please ? 

[0001-BUG-MEDIUM-cli-Never-wait-for-more-data-on-client-sh.patch.txt](https://github.com/haproxy/haproxy/files/7887233/0001-BUG-MEDIUM-cli-Never-wait-for-more-data-on-client-sh.patch.txt)

If it fixes your issue, it means your script sends partial commands or any junk data before closing. A good workaround is to fix the script.

**Comment by hujiko** [comment]
Hey @capflam,

thanks a lot for the patch.
I applied it, but the issue is still happening for me.
The socket stopped working ~14 Minutes after the new process started.

> If the script requesting the master CLI sends a partial command (or any data except newline or semicolon) before closing

This should not be, what the pyhton script does, but of course I can share the bits of it, that connect to the master socket:

```python
def process_ids():
    masterSocket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM);
    masterSocket.connect('/var/run/haproxy-master.sock');

    masterSocket.send("show proc\n".encode());

    proc_resp = masterSocket.recv(1024).decode().splitlines();
    raw_procs = [x for x in proc_resp if ("worker" in x and x[0] != '#')];
    pids = list(map(lambda x: x.split(' ')[0], raw_procs))
    return pids

def get_haproxy_report(pid):
    masterSocket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM);
    masterSocket.connect('/var/run/haproxy-master.sock');

    masterSocket.send(("@!%s show stat\n" % pid).encode());

    stats_resp = masterSocket.recv(9999999).decode().lstrip('# ');
    if stats_resp == "Can't connect to the target CLI!":
        return []
    return csv.DictReader(stats_resp.splitlines())
```

Unless the script receives an external termination signal liek SIGTERM / SIGKILL (which is does not), I can't see, how this would lead to sending partial commands.

In the end, this script runs in an endless loop with a sleep to every minute poll new data.
So I guess, if its the script, it must be a race condition, otherwise I would expect it to break exactly after 10 rounds (which would be exactly 10 Minutes)

**Comment by wtarreau** [comment]
Hi,

I've just found 3 other issues on the master CLI, I don't think they are directly related to what you're observing but I don't know their extent yet, and maybe they could be affected by timing variations. The first two ones seem to only affect pipelined commands and easily result in endless loops. The 3rd one can cause an endless loop on an empty buffer, but for me it doesn't quit the function and is stopped by the watchdog, so it's possible that it's different in your case. At least I need to sort this out before we can analyse further.

**Comment by wtarreau** [comment]
Finally knocked them down. They only affect long series of commands, which your script doesn't use so they shouldn't have any effect in your case :-/  No other idea at this point. If you want to try them anyway on top of Christophe's, here they are.

[0003-BUG-MINOR-stream-make-the-call_rate-only-count-the-n.patch.txt](https://github.com/haproxy/haproxy/files/7907519/0003-BUG-MINOR-stream-make-the-call_rate-only-count-the-n.patch.txt)
[0002-BUG-MEDIUM-mcli-always-realign-wrapping-buffers-befo.patch.txt](https://github.com/haproxy/haproxy/files/7907520/0002-BUG-MEDIUM-mcli-always-realign-wrapping-buffers-befo.patch.txt)
[0001-BUG-MEDIUM-mcli-do-not-try-to-parse-empty-buffers.patch.txt](https://github.com/haproxy/haproxy/files/7907521/0001-BUG-MEDIUM-mcli-do-not-try-to-parse-empty-buffers.patch.txt)

**Comment by hujiko** [comment]
Hey,

sorry for the late response.
I applied the three patches additionally, but (as you predicted) still have the same behavior.

**Comment by wlallemand** [comment]
It's kind of difficult to know if this is the master which is blocked, or the worker.

Since you configured a standard stats socket for the worker, are you able to connect to this one when the master CLI is blocked?

If that's the case, could you issue a "show sess all" and a "show fd" on this one? The interesting parts are the blocks with the "GLOBAL" frontend.

[0001-WIP-MINOR-mworker-cli-activate-some-debug-commands-f.patch.txt](https://github.com/haproxy/haproxy/files/7926445/0001-WIP-MINOR-mworker-cli-activate-some-debug-commands-f.patch.txt)

Attached is a patch which allows to do the same on the master CLI, you should be able to track the sessions (with "show sess all") of the "MASTER" frontend before it starts to block. The session on the master are supposed to disappear immediatly once the response is issued. If that's not the case, it will reach the maxconn of 10 and won't be able to accept new connections.

Also, are you seeing any anormal CPU usage on the master?

**Comment by capflam** [comment]
@hujiko, any news ? FYI, 2.5.4 was released

**Comment by hujiko** [comment]
@capflam no, sadly no good news.
I yesterday tried 2.5.4 from the Ubuntu PPA, but I still see the same Issue.
After around 30 Minutes the socket was stuck again.

Not sure if it helps, but I can add some graphs on the "load" of that instance.
It does around 10 TCP-Connections per second and a few http req/s (this HAProxy does some TCP and some HTTP listens).
The red marker indicates when the HAproxy reload happened, that activated HAProxy 2.5.4-1ppa1~focal.

<img width="1258" alt="HAProxy session rate and request rate" src="https://user-images.githubusercontent.com/946442/156371824-7f1f59ab-bf1e-4238-8944-6ac81be2a7c2.png">


Also we are now tracking some metrics regarding the OS network stack:
<img width="1407" alt="OS network metrics" src="https://user-images.githubusercontent.com/946442/156372005-73cc0575-adfc-4fc1-ae27-796b00124aec.png">

**Comment by capflam** [comment]
Thanks for the info. Could you apply the William's patch ? This way, it will be possible to use `show sess all` and `show fd` commands on the master socket. These both commands, executed on the master process, could help to find connection leak.

**Comment by hujiko** [comment]
Ah sorry, I must have somehow missed that post.

So what I did:

I switched to the upstream version 2.5.4
```
HAProxy version 2.5.4-1ppa1~focal 2022/02/25 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.4.html
Running on: Linux 5.4.0-80-generic #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021 x86_64
```

Then I waited for the master socket to hang again.

Apparently I was still able to connect to the worker socket. I had to change the access level form `user` to `admin`, to be able to execute the commands, but that's fine for testing.

So it is the master process, that is hanging and not forwarding stuff to the worker any more.

I now have the dumps from `show sess all` and `show fd`.
I would like to not publish them here, but rather send them directly to you.

Would it be fine, if I send an email to both of you with the files as attachments?
The mail addresses I would take from your GitHub-Commits, if that's fine for you.

**Comment by capflam** [comment]
Thanks to the (offline) help of @hujiko, we finally found the bug. It should be fixed now.

**Comment by hujiko** [comment]
Awesome, thank you very much

**Comment by capflam** [comment]
I'm reopening the issue. We will keep it open waiting for all backports.
